proofs/signatures/er.plf

```diff
@@ -0,0 +1,100 @@
+; Depends on sat.plf
+
+; This file exists to support the **Definition** rule in the paper:
+;  "Extended Resolution Simulates DRAT"
+; which can be found at http://www.cs.utexas.edu/~marijn/publications/ijcar18.pdf
+;
+; The core idea of extended resolution is that given **any** formula f
+; involving the variables from some SAT problem, one can introduce the
+; constraint
+;
+;    new <=> f
+;
+; without changing satisfiability, where "new" is a fresh variable.
+;
+; The paper uses the extension rule:
+;
+;    new <=> old v (~l_1 ^ ~l_2 ^ ... ^ ~l_n)
+;
+; which translates into the clauses:
+;
+;                      new v l_1 v l_2 v ... v l_n
+;                      new v ~old
+;     for each i <= n: l_i v ~new v old
+;
+; This translation could be proven using natural deduction, but instead we make
+; it an axiom for faster checking. It would be cool to make is a theorem, not
+; an axiom!
+
+(declare cnfc_unroll_towards_bottom
+         (! first clause
+            (! rest cnf
+               (! pf (cnf_holds (cnfc first rest))
+                  (! rest_of_pf (! r1 (holds first) (! r2 (cnf_holds rest) (holds cln)))
+                    (holds cln))))))
+
+; Takes a `list` of literals and a clause, `suffix`, adds the suffix to each
+; literal and returns a list of clauses as a `cnf`.
+(program clause_add_suffix_all ((list clause) (suffix clause)) cnf
+         (match list
+                (cln cnfn)
+                ((clc l rest) (cnfc (clc l suffix)
+                                    (clause_add_suffix_all rest suffix)))))
+
+; A definition witness value for:
+;              new <=> old v (~others_1 ^ ~others_2 ^ ... ^ ~others_n)
+; It witnesses the fact that new was fresh when it was defined by the above.
+; Thus it witnesses that the above is TAUT.
+(declare rat_elimination_def (! new var (! old lit (! others clause type))))
```

Comment:
This is a good question. The idea of extended resolution is that all statements of the form

    fresh <=> function of old variables

can be introduced without modifying satisfiability. Since a RAT-elimination definition has this form, it is justified by the idea of "pure" (would you permit to say _general_?) extended resolution.

However, RAT-elimination definitions are only a subset of the definitions justified by ER. They're a subset that was designed (a) to allow for the elimination of proper RATs and (b) to be easily convertible into clauses. These properties make them useful for converting from DRAT proofs to ER proofs.

So, in conclusion, this kind of definition is related to RATs in that its useful for eliminating them, and is also justified by the general idea of ER.

Reply:
I have a comment at the top of the file that meant to get at this, but I think that comment needs improvement. I'll try to fix it up.

Reply:
This is clearer now, thanks. However, I think that the signature itself will be clearer if it doesn't have lrat/drat notions in it (unless, of course, it is necessary [or, at least, useful] to know that a `definition` originates from a RAT elimination step in the current context).